***Carga y aproximación a la base de datos***

In [ ]:
#importamos las librerías que necesitamos 
import pandas as pd
import numpy as np

In [ ]:
#cargamos nuestras bases de datos
df1 = pd.read_csv('/content/p_data.csv',delimiter=';')
df2 = pd.read_csv('/content/w_data.csv',delimiter=';')

In [ ]:
#vemos los primeros registros de la base 
df1.head()

Age      Sex      ID    Race         Country   Education  Education-Num  \
0  39.0     Male   49084   White   United-States   Bachelors           13.0   
1  50.0     Male   68553   White   United-States   Bachelors           13.0   
2  38.0     Male  143654   White   United-States     HS-grad            9.0   
3  53.0     Male   41510   Black   United-States        11th            7.0   
4  28.0   Female  195025   Black            Cuba   Bachelors           13.0   

        Marital Status    Relationship  
0        Never-married   Not-in-family  
1   Married-civ-spouse         Husband  
2             Divorced   Not-in-family  
3   Married-civ-spouse         Husband  
4   Married-civ-spouse            Wife

In [ ]:
df2.head()

Workclass    fnlwgt  Capital Gain  Capital Loss  Hours per week  \
0          State-gov   77516.0        2174.0           0.0            40.0   
1   Self-emp-not-inc   83311.0           0.0           0.0            13.0   
2            Private  215646.0           0.0           0.0            40.0   
3            Private  234721.0           0.0           0.0            40.0   
4            Private  338409.0           0.0           0.0            40.0   

       ID  Target  
0   49084   <=50K  
1   68553   <=50K  
2  143654   <=50K  
3   41510   <=50K  
4  195025   <=50K

In [ ]:
#vemos cómo se componen ambos dataset 
print('El df1 se compone de', df1.shape[0], 'filas y de', df1.shape[1], 'columnas')

El df1 se compone de 48842 filas y de 9 columnas


In [ ]:
print('El df1 se compone de', df2.shape[0], 'filas y de', df2.shape[1], 'columnas')

El df1 se compone de 48843 filas y de 7 columnas


Vemos que el segundo dataframe tiene un registro más y dos columnas menos. 

***Unimos las dos tablas***

In [ ]:
#eliminamos los duplicados de acuerdo a la columna id del df1 
df1.drop_duplicates('ID')

Age      Sex          ID                 Race         Country  \
0      39.0     Male       49084                White   United-States   
1      50.0     Male       68553                White   United-States   
2      38.0     Male      143654                White   United-States   
3      53.0     Male       41510                Black   United-States   
4      28.0   Female      195025                Black            Cuba   
...     ...      ...         ...                  ...             ...   
48837  33.0     Male   597186815                White   United-States   
48838  39.0   Female   735617153                White   United-States   
48839  64.0     Male  1131807446                Black   United-States   
48840  38.0     Male   781422087                White   United-States   
48841  44.0     Male   161771296   Asian-Pac-Islander   United-States   

        Education  Education-Num       Marital Status     Relationship  
0       Bachelors           13.0        Never-married    Not-in-family  
1       Bachelors           13.0   Married-civ-spouse          Husband  
2         HS-grad            9.0             Divorced    Not-in-family  
3            11th            7.0   Married-civ-spouse          Husband  
4       Bachelors           13.0   Married-civ-spouse             Wife  
...           ...            ...                  ...              ...  
48837   Bachelors           13.0        Never-married        Own-child  
48838   Bachelors           13.0             Divorced    Not-in-family  
48839     HS-grad            9.0              Widowed   Other-relative  
48840   Bachelors           13.0   Married-civ-spouse          Husband  
48841   Bachelors           13.0             Divorced        Own-child  

[48842 rows x 9 columns]

In [ ]:
#eliminamos los duplicados de acuerdo a la columna id del df2 
df2.drop_duplicates('ID')

Workclass    fnlwgt  Capital Gain  Capital Loss  \
0              State-gov   77516.0        2174.0           0.0   
1       Self-emp-not-inc   83311.0           0.0           0.0   
2                Private  215646.0           0.0           0.0   
3                Private  234721.0           0.0           0.0   
4                Private  338409.0           0.0           0.0   
...                  ...       ...           ...           ...   
48838            Private  215419.0           0.0           0.0   
48839                  ?  321403.0           0.0           0.0   
48840            Private  374983.0           0.0           0.0   
48841            Private   83891.0        5455.0           0.0   
48842       Self-emp-inc  182148.0           0.0           0.0   

       Hours per week          ID   Target  
0                40.0       49084    <=50K  
1                13.0       68553    <=50K  
2                40.0      143654    <=50K  
3                40.0       41510    <=50K  
4                40.0      195025    <=50K  
...               ...         ...      ...  
48838            36.0   735617153   <=50K.  
48839            40.0  1131807446   <=50K.  
48840            50.0   781422087   <=50K.  
48841            40.0   161771296   <=50K.  
48842            60.0  1344995623    >50K.  

[48843 rows x 7 columns]

In [ ]:
#realizamos la unión de los dos dataframes con un right merge ya que el segundo df tiene un registro más que el primero 
df_complete = pd.merge(df1, df2, on = 'ID', how='right', indicator=True)
print(f'El dataframe completo tiene {len(df_complete)} filas y {df_complete.shape[1]} columnas')

El dataframe completo tiene 48843 filas y 16 columnas


In [ ]:
df_complete.head()

Age      Sex      ID    Race         Country   Education  Education-Num  \
0  39.0     Male   49084   White   United-States   Bachelors           13.0   
1  50.0     Male   68553   White   United-States   Bachelors           13.0   
2  38.0     Male  143654   White   United-States     HS-grad            9.0   
3  53.0     Male   41510   Black   United-States        11th            7.0   
4  28.0   Female  195025   Black            Cuba   Bachelors           13.0   

        Marital Status    Relationship          Workclass    fnlwgt  \
0        Never-married   Not-in-family          State-gov   77516.0   
1   Married-civ-spouse         Husband   Self-emp-not-inc   83311.0   
2             Divorced   Not-in-family            Private  215646.0   
3   Married-civ-spouse         Husband            Private  234721.0   
4   Married-civ-spouse            Wife            Private  338409.0   

   Capital Gain  Capital Loss  Hours per week  Target _merge  
0        2174.0           0.0            40.0   <=50K   both  
1           0.0           0.0            13.0   <=50K   both  
2           0.0           0.0            40.0   <=50K   both  
3           0.0           0.0            40.0   <=50K   both  
4           0.0           0.0            40.0   <=50K   both

***Data Wrangling***

Establecemos el ID como índice y lo ordenamos de menor a mayor

In [ ]:
df_complete.set_index('ID', inplace=True)

In [ ]:
df_complete = df_complete.sort_index()

In [ ]:
df_complete

Age      Sex    Race         Country   Education  Education-Num  \
ID                                                                             
3080        36.0     Male   White   United-States     Masters           14.0   
7286        62.0     Male   White   United-States     HS-grad            9.0   
7496        31.0   Female   White   United-States     Masters           14.0   
9929        44.0   Female   White   United-States     HS-grad            9.0   
17968       25.0   Female   White   United-States     Masters           14.0   
...          ...      ...     ...             ...         ...            ...   
3199610067  31.0     Male   White   United-States   Bachelors           13.0   
3200552514  19.0   Female   White   United-States     HS-grad            9.0   
3200755724  28.0     Male   White   United-States     HS-grad            9.0   
3225489111  43.0     Male   White          Mexico     HS-grad            9.0   
3229272340  72.0     Male   White   United-States     HS-grad            9.0   

                 Marital Status     Relationship     Workclass    fnlwgt  \
ID                                                                         
3080         Married-civ-spouse          Husband       Private  305319.0   
7286         Married-civ-spouse          Husband     Local-gov  114060.0   
7496              Never-married    Not-in-family       Private   45781.0   
9929                   Divorced   Other-relative       Private  103397.0   
17968             Never-married        Own-child       Private  179953.0   
...                         ...              ...           ...       ...   
3199610067            Separated    Not-in-family   Federal-gov  381645.0   
3200552514        Never-married        Own-child       Private   57145.0   
3200755724   Married-civ-spouse          Husband       Private   66095.0   
3225489111   Married-civ-spouse          Husband       Private  260761.0   
3229272340   Married-civ-spouse          Husband             ?  129912.0   

            Capital Gain  Capital Loss  Hours per week   Target _merge  
ID                                                                      
3080                 0.0           0.0            50.0    >50K.   both  
7286                 0.0           0.0            40.0    <=50K   both  
7496             14084.0           0.0            50.0     >50K   both  
9929                 0.0           0.0            40.0   <=50K.   both  
17968             2597.0           0.0            31.0    <=50K   both  
...                  ...           ...             ...      ...    ...  
3199610067           0.0           0.0            40.0    <=50K   both  
3200552514           0.0           0.0            25.0    <=50K   both  
3200755724           0.0           0.0            40.0    <=50K   both  
3225489111           0.0           0.0            40.0    <=50K   both  
3229272340           0.0           0.0            25.0    <=50K   both  

[48843 rows x 15 columns]

In [ ]:
#transformamos los nombres de las columnas a minuscula 
df_complete = df_complete.rename(columns={'Age': 'age', 'Sex': 'sex', 'Race': 'race', 'Country': 'country', 'Education': 'education', 'Target': 'target',
                                          'Education-Num': 'education-Num', 'Marital Status' : 'marital status','Relationship' : 'relationship',
                                          'Workclass': 'workclass','Capital Gain': 'capital gain', 'Capital Loss': 'capital loss', 'Hours per week': 'hours per week', '_merge': 'merge'})
df_complete

age      sex    race         country   education  education-Num  \
ID                                                                             
3080        36.0     Male   White   United-States     Masters           14.0   
7286        62.0     Male   White   United-States     HS-grad            9.0   
7496        31.0   Female   White   United-States     Masters           14.0   
9929        44.0   Female   White   United-States     HS-grad            9.0   
17968       25.0   Female   White   United-States     Masters           14.0   
...          ...      ...     ...             ...         ...            ...   
3199610067  31.0     Male   White   United-States   Bachelors           13.0   
3200552514  19.0   Female   White   United-States     HS-grad            9.0   
3200755724  28.0     Male   White   United-States     HS-grad            9.0   
3225489111  43.0     Male   White          Mexico     HS-grad            9.0   
3229272340  72.0     Male   White   United-States     HS-grad            9.0   

                 marital status     relationship     workclass    fnlwgt  \
ID                                                                         
3080         Married-civ-spouse          Husband       Private  305319.0   
7286         Married-civ-spouse          Husband     Local-gov  114060.0   
7496              Never-married    Not-in-family       Private   45781.0   
9929                   Divorced   Other-relative       Private  103397.0   
17968             Never-married        Own-child       Private  179953.0   
...                         ...              ...           ...       ...   
3199610067            Separated    Not-in-family   Federal-gov  381645.0   
3200552514        Never-married        Own-child       Private   57145.0   
3200755724   Married-civ-spouse          Husband       Private   66095.0   
3225489111   Married-civ-spouse          Husband       Private  260761.0   
3229272340   Married-civ-spouse          Husband             ?  129912.0   

            capital gain  capital loss  hours per week   target merge  
ID                                                                     
3080                 0.0           0.0            50.0    >50K.  both  
7286                 0.0           0.0            40.0    <=50K  both  
7496             14084.0           0.0            50.0     >50K  both  
9929                 0.0           0.0            40.0   <=50K.  both  
17968             2597.0           0.0            31.0    <=50K  both  
...                  ...           ...             ...      ...   ...  
3199610067           0.0           0.0            40.0    <=50K  both  
3200552514           0.0           0.0            25.0    <=50K  both  
3200755724           0.0           0.0            40.0    <=50K  both  
3225489111           0.0           0.0            40.0    <=50K  both  
3229272340           0.0           0.0            25.0    <=50K  both  

[48843 rows x 15 columns]

Para comparar cuales eran los registros que compartían el mismo ID, en el merge creamos una nueva columna que devuelve si el registro están en ambos df o left/right. 

In [ ]:
not_in_value = df_complete[df_complete['merge'] =='right_only']
not_in_value



age  sex race country education  education-Num marital status  \
ID                                                                          
1344995623  NaN  NaN  NaN     NaN       NaN            NaN            NaN   

           relationship      workclass    fnlwgt  capital gain  capital loss  \
ID                                                                             
1344995623          NaN   Self-emp-inc  182148.0           0.0           0.0   

            hours per week  target       merge  
ID                                              
1344995623            60.0   >50K.  right_only

Como vemos el ID 1344995623 es el que no tienen en común los dos df. 

Transformamos los target a dos valores únicamente (‘<=50k’ y ‘>50k’) para poder analizarlos posteriormente.

In [ ]:
#vemos cuáles son los distintos tipos de target que existen 
df_complete["target"].unique()

array([' >50K.', ' <=50K', ' >50K', ' <=50K.', nan], dtype=object)

In [ ]:
df_complete['target'].replace([(' >50K.'), (' >50K')], '>50k', inplace=True)
df_complete['target'].replace([(' <=50K'), (' <=50K.')], '<=50k', inplace=True)

In [ ]:
#comprobamos que los valores que quedaron fueron esos 
df_complete['target'].unique()

array(['>50k', '<=50k', nan], dtype=object)

In [ ]:
#vemos cuántos registros quedaron con la separación de target que realizamos 
df_complete.groupby(by='target').count()

age    sex   race  country  education  education-Num  \
target                                                           
<=50k   37155  37155  37155    37155      37155          37155   
>50k    11686  11686  11686    11686      11686          11686   

        marital status  relationship  workclass  fnlwgt  capital gain  \
target                                                                  
<=50k            37155         37155      37155   37155         37155   
>50k             11686         11686      11687   11687         11687   

        capital loss  hours per week  merge  
target                                       
<=50k          37155           37155  37155  
>50k           11687           11687  11687

Transormamos las columnas de sexo, raza y educación para poder analizarla posteriormente. 

In [ ]:
#comenzamos viendo cuántos valores distintos existen en la columna race
df_complete['race'].unique()

array([' White', ' Other', nan, ' Black', ' Asian-Pac-Islander',
       ' Amer-Indian-Eskimo'], dtype=object)

In [ ]:
#modificamos los valores para que estén en minúscula y sin espacios en blanco
df_complete = df_complete.replace({'race': {' White':'white',
                                   ' Other':'other',
                                   ' Black':'black',
                                   ' Asian-Pac-Islander': 'asian-pac-islander',
                                   ' Amer-Indian-Eskimo': 'amer-indian-eskimo'}})

In [ ]:
#comprobamos si los valores se modificaron 
df_complete['race'].unique()

array(['white', 'other', nan, 'black', 'asian-pac-islander',
       'amer-indian-eskimo'], dtype=object)

In [ ]:
#continuamos modificando los valores de sexo, primero vemos cuáles son los valores existentes
df_complete['sex'].unique()

array([' Male', ' Female', nan], dtype=object)

In [ ]:
#no existe variablilidad en la columna de sexo, no es necesario unificar. Unicamente borraremos los espacios en blanco y tranformalo a minuscula
df_complete = df_complete.replace({'sex': {
                                   ' Male':'male',
                                   ' Female': 'female'}})

In [ ]:
#comprobamos que los cambios se realizaron correctamente
df_complete['sex'].unique()

array(['male', 'female', nan], dtype=object)

In [ ]:
#vemos los valores que están en la columna educación 
df_complete['education'].unique()

array([' Masters', ' HS-grad', ' Prof-school', ' Bachelors', nan, ' 11th',
       ' Some-college', ' 9th', ' Assoc-acdm', ' Assoc-voc', ' 10th',
       ' 7th-8th', ' Doctorate', ' 1st-4th', ' 5th-6th', ' Preschool',
       ' 12th'], dtype=object)

In [ ]:
#son varios los valores a los que hay quitarle el espacio y pasarlo a miniscula
df_complete['education'] = df_complete['education'].str.lower()
df_complete['education'] = df_complete['education'].str.strip()

In [ ]:
#comprobamos haber modificado los valores correctamente
df_complete['education'].unique()

array(['masters', 'hs-grad', 'prof-school', 'bachelors', nan, '11th',
       'some-college', '9th', 'assoc-acdm', 'assoc-voc', '10th',
       '7th-8th', 'doctorate', '1st-4th', '5th-6th', 'preschool', '12th'],
      dtype=object)

Estando ya transformadas todas las columnas que necesitamos para el análisis (race, sex, education y target) procedemos a realiazarlo. 

Dentro del análisis nos planteamos preguntas tales como:
- ¿Cuál es el porcentaje de mujeres con un salario mayor a 50K?¿Qué diferencia existe con los hombres?
- ¿En su mayoría la raza blanca qué salario tiene?
- ¿En su mayoría la raza amer-indian-eskimo qué salario tiene?
- ¿Todos los que tienen un máster tienen salarios igual o mayores a 50K?

In [ ]:
#eliminamos los valores nulos 
df_complete.dropna(inplace=True)
df_complete.shape

(48841, 15)

In [ ]:
#creamos las variables que necesitamos para el analisis según los sexos 
high_salary_women = df_complete[(df_complete['sex'] == 'female') & (df_complete['target'] == '>50k')].count()
low_salary_women = df_complete[(df_complete['sex']== 'femele') & (df_complete['target'] == '<=50k')].count()
high_salary_men = df_complete[(df_complete['sex'] == 'male') & (df_complete['target'] == '>50k')].count()

In [ ]:
print(f'El porcentaje de mujeres que tienen un salario mayor a 50K es de {round(high_salary_women[0] / df_complete.count()[0] * 100,2)} % ')
print(f' mientras el porcentaje de hombres que tienen un salario mayor a 50K es de {round((high_salary_men[0] / df_complete.count()[0]) * 100,2)} %')

El porcentaje de mujeres que tienen un salario mayor a 50K es de 3.62 % 
 mientras el porcentaje de hombres que tienen un salario mayor a 50K es de 20.3 %
